source/material_model/equation_of_state/thermodynamic_table_lookup.cc

```diff
@@ -0,0 +1,611 @@
+        prm.declare_entry ("Bilinear interpolation", "true",
+                           Patterns::Bool (),
+                           "Whether to use bilinear interpolation to compute "
+                           "material properties (slower but more accurate). ");
+        prm.declare_entry ("Latent heat", "false",
+                           Patterns::Bool (),
+                           "Whether to include latent heat effects in the "
+                           "calculation of thermal expansivity and specific heat. "
+                           "If true, ASPECT follows the approach of Nakagawa et al. 2009, "
+                           "using pressure and temperature derivatives of the enthalpy "
+                           "to calculate the thermal expansivity and specific heat. "
+                           "If false, ASPECT uses the thermal expansivity and "
+                           "specific heat values from the material properties table.");
+        prm.declare_entry ("Maximum latent heat substeps", "1",
+                           Patterns::Integer (1),
+                           "The maximum number of substeps over the temperature pressure range "
+                           "to calculate the averaged enthalpy gradient over a cell.");
+      }
+
+
+
+      template <int dim>
+      void
+      ThermodynamicTableLookup<dim>::parse_parameters (ParameterHandler &prm)
+      {
+        data_directory               = Utilities::expand_ASPECT_SOURCE_DIR(prm.get ("Data directory"));
+        material_file_names          = Utilities::split_string_list(prm.get ("Material file names"));
+        n_material_lookups           = material_file_names.size();
+
+        derivatives_file_names       = Utilities::split_string_list(prm.get ("Derivatives file names"));
+        use_bilinear_interpolation   = prm.get_bool ("Bilinear interpolation");
+        latent_heat                  = prm.get_bool ("Latent heat");
+        max_latent_heat_substeps     = prm.get_integer ("Maximum latent heat substeps");
+
+        has_background               = prm.get_bool ("Background material");
+        first_composition_index      = prm.get_integer ("Index of first mass fraction compositional field");
+
+        if (prm.get ("Material file format") == "perplex")
+          material_file_format       = perplex;
+        else if (prm.get ("Material file format") == "hefesto")
+          material_file_format       = hefesto;
+        else
+          AssertThrow (false, ExcNotImplemented());
+
+        // Do some error checking
+
+        // First, do a fairly lax check on the number of material files.
+        // If the following comparison evaluates to be an equality,
+        // it implies that all compositional fields after
+        // "first_composition_index" correspond to mass fractions
+
+        if (has_background)
+          {
+            AssertThrow ((material_file_names.size() <= this->n_compositional_fields() - first_composition_index + 1),
```

Comment:
As mentioned in the previous review, I think I would just remove the first_composition_index here and everywhere and assume it is zero. The option was not available before (one could have changed the hard-coded number, but it wasn't an option in the input file), so we can easily remove it and instead later on add an option to define the compositional fields that represent mass fractions.

Reply:
Hmmm, ok, you make a good point. Maybe this is a good opportunity to implement the compositional type labelling...